### PR TITLE
Allow variable pool size on BaseProtocol / Protocol

### PR DESCRIPTION
**Pull request by jhatch28**
Due to the blanket increase in session pool size to 500, we were seeing memory issues. 
This is a more targeted solution that adds support for larger pool sizes only for service accounts. 

I think the memory issues were two-fold:
- exchangelib creates a bunch of `requests.session.Session` objects, each with a single TCP connection that remains open once created
- exchangelib creates a bunch of multiprocesses using the `ThreadPool`

Tested this locally in a debugger and stepped through each part to verify it has the proper behavior

### CODE REVIEW

**Review by pengfeiye — APPROVED** [review]
LGTM, but not super familiar with Exchangelib if we can get another pair of eyes on this that be great.

File: exchangelib/protocol.py
Comment (singingwolfboy) [inline exchangelib/protocol.py]: Can we avoid introducing Nylas-specific concepts into exchangelib? At some point, we will definitely want to switch away from our fork and use the upstream version -- and when that happens, it will be more difficult to make the switch if we have to pull out Nylas-specific concepts.

Maybe we could just set the `SESSION_POOLSIZE` value on the `Protocol` objects for the accounts we want to change?

**Review by singingwolfboy — CHANGES_REQUESTED** [review]
I'm not convinced that this change is necessary at all. Can we just set the `SESSION_POOLSIZE` value on the `Protocol` object when we construct it in the Nylas code?

I should also point out that the upstream version of exchangelib has actually removed the ThreadPool feature entirely in [version 3.2.0](https://github.com/ecederstrand/exchangelib/blob/master/CHANGELOG.md#320), and it sounds like this might be related to that feature. From the changelog:

> Remove use of `ThreadPool` objects. Threads were used to implement async HTTP requests, but were creating massive memory leaks. Async requests should be reimplemented using a real async HTTP request package, so this is just an emergency fix. This also lowers the default `Protocol.SESSION_POOLSIZE` to 1 because no internal code is running multi-threaded anymore.

Is there some way that we can do something similar, if we're dealing with memory issues?

File: exchangelib/protocol.py
Comment (jhatch28) [reply exchangelib/protocol.py]: I agree, but I don't see a way to set `SESSION_POOLSIZE` on a specific instance of `Protocol` if it sets up the session and thread pools in the `__init__`. Do you have a suggestion?

Comment (singingwolfboy) [reply exchangelib/protocol.py]: Do we have a way to access the Protocol object from the Nylas code? If so, we can just assign to `SESSION_POOLSIZE` on the object. There are no private variables in Python, everything is public.

Comment (jhatch28) [reply exchangelib/protocol.py]: Yes. But how do we set a class (not instance) attribute and guarantee it won't affect other instances created?

We could potentially move this session pool setup into a separate method that can take a `pool_size` kwarg that defaults to `SESSION_POOLSIZE` and call it in `__init__`. Then, when the instance is created for service accounts, we can call that method again from our code with the updated pool size to overwrite the existing pools. 

Another option is to switch out `is_service_account` with a `pool_size` kwarg

**Comment by ecederstrand** [comment]
I hope it's alright that I jump in here (I'm the upstream exchangelib maintainer).

`Protocol` instances are cached and re-used for multiple accounts on the same server (see the `CachingProtocol` class). Introducing per-account settings doesn't really make sense because the instance is per-server.

The fix to the memory leak is almost certainly to remove the usage of `ThreadPool`, as was done upstream: https://github.com/ecederstrand/exchangelib/commit/010c417e0094cd7730f2dd8bef7a0cef4f061885

Also, the `Account` class models an Exchange mailbox (a container for mail, calendar, notes etc), while a service account in Exchange is a special kind of user that may or may not have a mailbox. This is what the `Credentials` class in exchangelib models. Introducing the service account concept into the Account class is conceptually incorrect.

**Comment by jhatch28** [comment]
@ecederstrand we appreciate you jumping in! Sorry for the delayed reply, we're in a code freeze so this fix was de-prioritized since it won't get shipped for a couple weeks. 
We're planning to upgrade to your upstream version as soon as we can, so it's great to see that you have a fix already implemented for the memory leak we ran into. 

>Also, the Account class models an Exchange mailbox (a container for mail, calendar, notes etc), while a service account in Exchange is a special kind of user that may or may not have a mailbox. This is what the Credentials class in exchangelib models. Introducing the service account concept into the Account class is conceptually incorrect.

I completely agree with you here, and I think @singingwolfboy was making that point above as well. 

One remaining question I have: will `SESSION_POOLSIZE` still be a bottleneck if we have 10s-100s of accounts in the same process using the same service account credentials? Especially when each account will constantly have a `GetStreamingEvents` request active. 


P.S. thanks for all your hard work on this library!

**Comment by ecederstrand** [comment]
My experience with Exchange servers is that they are not happy with hundreds of TCP connections from the same user. Connections will be rate-limited, or the server may close connections or throw exceptions telling you to throttle. Exchange admins can disable that limit via policies on the server, but I'm not sure it's possible on O365 anymore, for example.

Upstream exchangelib doesn't support streaming events yet (I'm in contact with Christine Spang about porting your work) so I'm not familiar with how that works. E.g. does streaming events require a TCP connection open for each mailbox, or can they share connections?

**Comment by jhatch28** [comment]
You are exactly right - we had to do a lot of custom work on the client's Exchange Server to get 10k accounts running through impersonation with a single service account. However, that's something we can work with and our customers want, so Nylas needs to be able to support it. 

I didn't realize the streaming events wasn't supported upstream, but yes, it opens a TCP connection for up to 30 minutes. Each one needs its own connections IIRC, because you can only subscribe to a single SMTP Address. [MS Reference](https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/getstreamingevents-operation) in case you're interested.

**Comment by jhatch28** [comment]
_Finally_ revisiting this. What do you think of my suggested change to accept a `pool_size` parameter for now until we can use the upstream?

We can essentially just drop the change when we go to upstream, if the `SESSION_POOLSIZE` won't be the blocker. My opinion is that it _is_ still going to prevent us from opening more than 4 concurrent connections even in upstream, though. Am I wrong @ecederstrand ?

**Comment by ecederstrand** [comment]
Hi @jhatch28

Some thoughts on this:

In exchangelib, we use a Session and a TCP connection interchangeably. Internally, a Session object has a TCP connection pool that can have an arbitrary size. The connection pool is used when you use the Session in threaded environments. In exchangelib, we hand out Sessions to threads that need to connect, and end up never using more than 1 TCP connection from the Session connection pool. The reason is described in the comments for `BaseProtocol. CONNECTIONS_PER_SESSION`, but essentially, we want to make sure that authenticated connections are not shared between accounts.

The maximum number of TCP connections allowed for a given Exchange server is a function of the Exchange server (EWS endpoint) and the connecting user (credentials). exchangelib does not handle this fact very well. Protocol instances are cached with a cache key of `('service_endpoint', 'credentials')`. You can adjust `BaseProtocol.SESSION_POOLSIZE` just before connecting to a new server with a new set of credentials, but you have to be careful when using threads, of course.

I think the proposed change is fine as a workaround until this is handled better in upstream code.

**Comment by jhatch28** [comment]
@ecederstrand thanks for weighing in on this!

**Review by singingwolfboy — APPROVED** [review]

**Comment by ecederstrand** [comment]
Just FYI, I added a `max_connections` attribute to the `Configuration` class to handle this. A Configuration is usually created anyway, when you want to connect to a specific server with specific credentials, so an option to set the max poolsize fits nicely there. Here's the upstream commit: https://github.com/ecederstrand/exchangelib/commit/72d81e766a837f866a5b445f51be64830bebf68a

**Comment by jhatch28** [comment]
Thanks for the change and heads up @ecederstrand. That will be very nice for when we got onto the upstream version

**Pull request merged**
